### PR TITLE
Cv2 3435 alegre with presto for audio

### PR DESCRIPTION
**Pull request by DGaffney**
## Description

This ticket integrates alegre's presto usage into Check-API for dealing with callbacks from storage requests on alegre. As stated in the ticket, I think we can actually get away with a pretty narrow implementation: 

At the point that the webhook for storing data is executed, we don’t actually need to make a callback-based search query - we can use the search functionality nearly exactly as written today. When a new item is added in Check-API, we send a request to Alegre to store it, which then executes a callback against presto. Alegre receives the response from Presto, and at that point, we send a callback in turn from Alegre back to Check-API. At that point, when we receive the callback on Check-API from Alegre, we know that the data is now properly stored within Alegre - so we can just synchronously send a request to Alegre to do the lookup as it exists today. In fact, we should probably…. entirely rollback the portion of changes on Alegre for the GET requests (e.g. search-side) for now. The overwhelmingly common flow is for Alegre GETs to be used when we already know that the item is now safely stored, so in those rare cases where we aren’t sure, we should route through the new webhook setup instead of making everything be doubly-called-back. Make sense?

References: 3435

## How has this been tested?

I've updated and fixed the tests on alegre to assert an expected message packet to be sent to alegre, and I'm mapping that directly to the added webhook, and then patching directly into existing functionality. Should be pretty minor if any difference downstream of that.

## Things to pay attention to during code review

We will need to extensively test full service-to-service integrations just because this is involving so many different repos....

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [x] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

**Comment by computermacgyver** [comment]
There are functions such as `Bot::Alegre.get_items_with_similar_media` that expect Alegre to return similar results based on the a media file specified as a URL. My reading of https://github.com/meedan/alegre/pull/340 is that this no longer works unless the audio has been stored already in Alegre with that URL. This assumption does not hold when this function is called from Smooch Search (https://github.com/meedan/check-api/blob/c983b834a859c4052553b8cf102e13f79ba78577/app/models/concerns/smooch_search.rb#L167C5-L167C5) and likely also when it is called from Check Web search. 

Currently Alegre will temporarily hash and store an Audio file to perform the search, but this will no longer be the case after https://github.com/meedan/alegre/pull/340 is merged. We therefore need to identify such instances and either (a) create exceptions for audio to fail gracefully or (b) solve the more general case of how we'll use callbacks for tipline/check-web search

**Review by caiosba — COMMENTED** [review]
I'm going to review this once I'm able to run it locally, in order to make some tests.

File: app/models/bot/alegre.rb
Comment (computermacgyver) [inline app/models/bot/alegre.rb:504]: Do we need a line `parsed_response` after `end` to ensure `parsed_response` is the return value if non-callback functions? (I always get confused with this part of Ruby)

**Review by computermacgyver — COMMENTED** [review]
Looking great

File: app/models/bot/alegre.rb
Comment (caiosba) [inline app/models/bot/alegre.rb:131]: Let's also be sure that `!request.query_parameters['token'].blank?`

File: app/models/concerns/alegre_webhooks.rb
Comment (caiosba) [inline app/models/concerns/alegre_webhooks.rb:16]: Let's add a prefix to the Redis keys, just for better scoping/debugging if needed, like a namespace... so, instead of `doc_id`, something like `"alegre:webhook:#{doc_id}"`, for example.

File: app/models/bot/alegre.rb
Comment (caiosba) [reply app/models/bot/alegre.rb:504]: Let's please add a prefix to the Redis keys, just for better scoping/debugging if needed, like a namespace... so, instead of just `parsed_response.dig("body", "id")`, something like "alegre:webhook:#{parsed_response.dig("body", "id")}", for example.

Comment (caiosba) [reply app/models/bot/alegre.rb:504]: We do, @computermacgyver !

Comment (DGaffney) [reply app/models/bot/alegre.rb:504]: Duh..... ok I will address in morning!

File: app/models/concerns/alegre_webhooks.rb
Comment (DGaffney) [reply app/models/concerns/alegre_webhooks.rb:16]: works for me!

File: app/models/bot/alegre.rb
Comment (caiosba) [inline app/models/bot/alegre.rb:504]: Request: Please add a test for this case.

File: app/models/concerns/alegre_webhooks.rb
Comment (caiosba) [inline app/models/concerns/alegre_webhooks.rb:18]: This block is not tested, but I'll explain why in the next comment.

File: test/models/bot/alegre_3_test.rb
Comment (caiosba) [inline test/models/bot/alegre_3_test.rb:584]: If I'm reading it correctly, looks like this test is not actually doing anything. I suggest you move this test to https://github.com/meedan/check-api/blob/develop/test/controllers/webhooks_controller_test.rb and change it to do the following:

1. Create testing data and stub Alegre response, like you're already doing
2. Be sure that the Redis key is nil (you can use `assert_nil`)
3. Call the webhook endpoint by executing `post :index, params: { name: :alegre }, body: payload` in the test (you'll need to define `payload` similarly to how you are defining `request` as an `OpenStruct` currently) - there are examples in https://github.com/meedan/check-api/blob/develop/test/controllers/webhooks_controller_test.rb
4. Be sure that the key in Redis is not null and contains the expected value (you can use `assert_no_nil` or even `assert_equal`)
5. Use `travel_to` and be sure that the key was removed from Redis

**Review by caiosba — CHANGES_REQUESTED** [review]
Thanks Devin! We're almost there! I added a couple of requests about the tests, but other than that, please also do:

- Rebase this branch against latest `develop`
- Add `alegre_token: dev` to `config/config.yml.example`

File: test/test_helper.rb
Comment (caiosba) [inline test/test_helper.rb:154]: Why you commented this line?

File: test/controllers/webhooks_controller_test.rb
Comment (caiosba) [inline test/controllers/webhooks_controller_test.rb:236]: Please delete the `zzz` :zzz: from the test name lol

File: app/models/bot/alegre.rb
Comment (caiosba) [inline app/models/bot/alegre.rb:133]: Please move this to the `AlegreWebhooks` concern.

**Review by computermacgyver — APPROVED** [review]
Great work.

**Review by caiosba — APPROVED** [review]
I pushed a few cosmetic changes, simplified one test and added one more test. Good to go after Alegre is merged!

**Pull request merged**
